### DISCUSSION TITLE
Where are relations stored?

### DISCUSSION THREAD

**Discussion by srowe0091** [prisma/prisma#2756] [Q&A] [answered]
Hello all, I'm very new to prisma, so I'm trying to understand the concept of how Prisma handles relations. From my understanding, when we went to create relation from one model to another model, we use the `connect` and that creates the relation one-to-(one OR many) but where is that relation stored? I see in the docs that those kinds of relations are not stored in database. So does that mean those relations are stored locally? What happens if the application/database needs to be moved to another server/service, are those relations are essentially lost?

**Comment by ryands17** [comment, marked as answer]
Hey @mrowe009 :wave: 

> Hello all, I'm very new to prisma, so I'm trying to understand the concept of how Prisma handles relations. From my understanding, when we went to create relation from one model to another model, we use the connect and that creates the relation one-to-(one OR many) but where is that relation stored?

This relation is stored directly in the database. When you create a `schema.prisma` file and run the migrations, it creates a relation in the database itself. For e.g.

```prisma
model Post {
  id        String   @default(cuid()) @id
  published Boolean
  title     String
  content   String?
  authorId  Int?
  author    User?    @relation(fields: [authorId], references: [id])
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
}

model User {
  id        Int      @default(autoincrement()) @id
  email     String   @unique
  password  String
  name      String?
  posts     Post[]
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
}
```

The `authorId` in the above is a relation to the User table and that's directly created in the DB.

> I see in the docs that those kinds of relations are not stored in database. So does that mean those relations are stored locally?

Could you point to the docs where the above point is mentioned?

> What happens if the application/database needs to be moved to another server/service, are those relations are essentially lost?

Prisma creates migrations in a specific folder, so if you need to move the database someplace else, you just need to run the migrations and it will work fine.

**Reply by srowe0091** [reply]
I understand the relations are created, but whenever you `connect` models together, that is what is not being stored. Here in the [docs](https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-schema/relations) in the overview section you will see this

> At a Prisma level, a relation is made up of:
> Two relation fields: author and posts. Relation fields are used to define connections between models at the Prisma level and do not exist in the database.

So that is what is confusing me because whenever i `connect` a relation, i do not see anything in the database (or maybe im looking in the wrong area), but I am still able to retrieve that connected data when executing queries. So if its not being stored in the database, then where is it being stored?

**Reply by ryands17** [reply]
As mentioned above, if we have a schema like this:

```prisma
model Post {
  id        String   @default(cuid()) @id
  published Boolean
  title     String
  content   String?
  authorId  Int?
  author    User?    @relation(fields: [authorId], references: [id])
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
}

model User {
  id        Int      @default(autoincrement()) @id
  email     String   @unique
  password  String
  name      String?
  posts     Post[]
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
}
```

And we create a Post from the User in the following manner:

```ts
  await prisma.user.create({
    data: {
      name: 'abc',
      email: 'email',
      password: 'password',
      posts: {
        create: {
          title: 'title',
          content: 'content',
          published: false,
        },
      },
    },
  })
```

The authoriId field is saved with the `id` field of the User in the Posts table (foreign key). On create, Prisma automatically adds the foreign key to the Posts table which we have mentioned in the above schema like this:

```prisma
authorId  Int?
author    User?    @relation(fields: [authorId], references: [id])
```

This is how Prisma knows how to form connections and save the required fields in the foreign key of the database with the help of the fields and references in the above relation.

So when you want to retrieve the connected data like this:

```ts
  await prisma.user.findOne({
    where: {id: 1},
    include: {
      posts: true
    }
  })
```

It will find all the posts where the `authorId` in the database has the value 1. This is internally done by Prisma, but it is exposed to you as `create` and `connect` while inserting relation data.

Let me know if that helps :)

**Reply by janpio** [reply]
@ryands17 The documentation could make this a bit more explicit, by saying that the relation scalar field is waht stores the information about what is connected to each other.

**Reply by srowe0091** [reply]
Okay, thank you very much for explaining that. Maybe im not seeing that relation in the database, but i'll try again. Thank you for answering this @ryands17 !

**Comment by ludralph** [comment]
Hi @srowe0091,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
